MXNet 1.6.0 CPU wheel incompatible with horovod

**Environment:**
1. Framework:  Tensorflow
2. Framework version: 2.3.0
3. Horovod version: latest
4. MPI version: N/A
5. CUDA version: N/A
6. NCCL version: N/A
7. Python version: 3.7
8. Spark / PySpark version: N/A
9. OS and version: ubuntu 18.04.1
10. GCC version: 4.8
11. CMake version: Not sure


I am installing horovodrun on CPU only mode using https://raw.githubusercontent.com/horovod/horovod/master/Dockerfile.cpu

However installation fails. Complete error log is in https://github.com/mangleddata/repro_horovodcpu/blob/main/err.log

Strangely, I was using DockerFile because I was running into the same issue without docker. Any idea on how to setup on a CPU only mode for distributed training ?

Looks like the error message here is:

```
Linking CXX shared library ../../../lib.linux-x86_64-3.7/horovod/mxnet/mpi_lib.cpython-37m-x86_64-linux-gnu.so
    cd /tmp/pip-install-7zdqpljv/horovod/build/temp.linux-x86_64-3.7/horovod/mxnet && /usr/bin/cmake -E cmake_link_script CMakeFiles/mxnet.dir/link.txt --verbose=1
    /usr/bin/c++ -fPIC  -DMXNET_USE_MKLDNN=0 -DMSHADOW_USE_CUDA=0 -D_GLIBCXX_USE_CXX11_ABI=1  -pthread -fPIC -Wall -ftree-vectorize -mf16c -mavx -mfma -DMSHADOW_USE_MKL=0 -DMSHADOW_USE_F16C=0 -O3 -g -DNDEBUG  -Wl,--version-script=/tmp/pip-install-7zdqpljv/horovod/horovod.lds -Wl,-Bsymbolic-functions -Wl,-z,relro,-z,now -shared -Wl,-soname,mpi_lib.cpython-37m-x86_64-linux-gnu.so -o ../../../lib.linux-x86_64-3.7/horovod/mxnet/mpi_lib.cpython-37m-x86_64-linux-gnu.so CMakeFiles/mxnet.dir/__/common/common.cc.o CMakeFiles/mxnet.dir/__/common/controller.cc.o CMakeFiles/mxnet.dir/__/common/fusion_buffer_manager.cc.o CMakeFiles/mxnet.dir/__/common/half.cc.o CMakeFiles/mxnet.dir/__/common/logging.cc.o CMakeFiles/mxnet.dir/__/common/message.cc.o CMakeFiles/mxnet.dir/__/common/operations.cc.o CMakeFiles/mxnet.dir/__/common/parameter_manager.cc.o CMakeFiles/mxnet.dir/__/common/response_cache.cc.o CMakeFiles/mxnet.dir/__/common/stall_inspector.cc.o CMakeFiles/mxnet.dir/__/common/thread_pool.cc.o CMakeFiles/mxnet.dir/__/common/timeline.cc.o CMakeFiles/mxnet.dir/__/common/tensor_queue.cc.o CMakeFiles/mxnet.dir/__/common/ops/collective_operations.cc.o CMakeFiles/mxnet.dir/__/common/ops/operation_manager.cc.o CMakeFiles/mxnet.dir/__/common/optim/bayesian_optimization.cc.o CMakeFiles/mxnet.dir/__/common/optim/gaussian_process.cc.o CMakeFiles/mxnet.dir/__/common/utils/env_parser.cc.o CMakeFiles/mxnet.dir/__/common/mpi/mpi_context.cc.o CMakeFiles/mxnet.dir/__/common/mpi/mpi_controller.cc.o CMakeFiles/mxnet.dir/__/common/ops/mpi_operations.cc.o CMakeFiles/mxnet.dir/__/common/ops/adasum/adasum_mpi.cc.o CMakeFiles/mxnet.dir/__/common/ops/adasum_mpi_operations.cc.o CMakeFiles/mxnet.dir/__/common/gloo/gloo_context.cc.o CMakeFiles/mxnet.dir/__/common/gloo/gloo_controller.cc.o CMakeFiles/mxnet.dir/__/common/gloo/http_store.cc.o CMakeFiles/mxnet.dir/__/common/gloo/memory_store.cc.o CMakeFiles/mxnet.dir/__/common/ops/gloo_operations.cc.o CMakeFiles/mxnet.dir/mpi_ops.cc.o CMakeFiles/mxnet.dir/tensor_util.cc.o CMakeFiles/mxnet.dir/cuda_util.cc.o CMakeFiles/mxnet.dir/adapter.cc.o  -L/usr/local/lib/python3.7/dist-packages/mxnet -Wl,-rpath,/usr/local/lib:/usr/local/lib/python3.7/dist-packages/mxnet /usr/local/lib/libmpi.so -lmxnet ../../third_party/gloo/gloo/libgloo.a /usr/local/lib/libmpi.so -lpthread
    /usr/local/lib/python3.7/dist-packages/mxnet/libmxnet.so: file not recognized: File format not recognized
    collect2: error: ld returned 1 exit status
    horovod/mxnet/CMakeFiles/mxnet.dir/build.make:904: recipe for target '../lib.linux-x86_64-3.7/horovod/mxnet/mpi_lib.cpython-37m-x86_64-linux-gnu.so' failed
```

@mangleddata one workaround if you only need TensorFlow is to change [this line](https://github.com/horovod/horovod/blob/master/Dockerfile.cpu#L60) to:

```
RUN HOROVOD_WITH_TENSORFLOW=1 HOROVOD_WITHOUT_PYTORCH=1 HOROVOD_WITHOUT_MXNET=1 \
```

@leezu, any thoughts on why this error would occur with MXNet 1.6.0 on CPU?  I noticed we don't have this version under CI for CPU images, which we should update.

--- a/Dockerfile.cpu
+++ b/Dockerfile.cpu
@@ -3,7 +3,7 @@ FROM ubuntu:18.04
 ENV TENSORFLOW_VERSION=2.3.0
 ENV PYTORCH_VERSION=1.6.0
 ENV TORCHVISION_VERSION=0.7.0
-ENV MXNET_VERSION=1.6.0
+ENV MXNET_VERSION=1.5.0
 
 # Python 3.7 is supported by Ubuntu Bionic out of the box
 ARG python=3.7
@@ -57,7 +57,7 @@ RUN mkdir /tmp/openmpi && \
     rm -rf /tmp/openmpi
 
 # Install Horovod
-RUN HOROVOD_WITH_TENSORFLOW=1 HOROVOD_WITH_PYTORCH=1 HOROVOD_WITH_MXNET=1 \
+RUN MAKEFLAGS="-j1" HOROVOD_WITH_TENSORFLOW=1 HOROVOD_WITH_PYTORCH=1 HOROVOD_WITH_MXNET=1 \
     pip install --no-cache-dir horovod
 
 # Install OpenSSH for MPI to communicate between containers

--- a/build-docker-images.sh
+++ b/build-docker-images.sh
@@ -19,7 +19,7 @@ function build_one()
     fi
 
     tag=horovod-build-py${py}-${device}:$(date +%Y%m%d-%H%M%S)
-    docker build -f Dockerfile.${device} -t ${tag} --build-arg python=${py} --no-cache .
+    docker build -f Dockerfile.${device} -t ${tag} --build-arg python=${py} .
     horovod_version=$(docker run --rm ${tag} pip show horovod | grep Version | awk '{print $2}')
     tensorflow_version=$(docker run --rm ${tag} pip show ${tensorflow_pkg} | grep Version | awk '{print $2}')
     pytorch_version=$(docker run --rm ${tag} pip show torch | grep Version | sed 's/+/ /g' | awk '{print $2}')
@@ -30,12 +30,12 @@ function build_one()
 }
 
 # clear upstream images, ok to fail if images do not exist
-docker rmi $(cat Dockerfile.cpu | grep FROM | awk '{print $2}') || true
-docker rmi $(cat Dockerfile.gpu | grep FROM | awk '{print $2}') || true
+# docker rmi $(cat Dockerfile.cpu | grep FROM | awk '{print $2}') || true
+# docker rmi $(cat Dockerfile.gpu | grep FROM | awk '{print $2}') || true
 
 # build for cpu and gpu
 build_one 3.7 cpu
-build_one 3.7 gpu
+# build_one 3.7 gpu
 
 # print recent images
 docker images horovod/horovod

